lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+
+    var prom;
+    if (initialEventId) {
+        debuglog("TimelineWindow: initialising for event " + initialEventId);
+        prom = this._client.getEventTimeline(this._room, initialEventId)
+            .then(function(tl) {
+                // make sure that our window includes the event
+                for (var i = 0; i < tl.getEvents().length; i++) {
+                    if (tl.getEvents()[i].getId() == initialEventId) {
+                        return [tl, i];
+                    }
+                }
+                throw new Error("getEventTimeline result didn't include requested event");
+            });
+    } else {
+        debuglog("TimelineWindow: initialising with live timeline");
+
+        // start with the most recent events
+        var tl = this._room.getLiveTimeline();
+        prom = q([tl, tl.getEvents().length]);
+    }
+
+    prom = prom.then(function(v) {
+        var tl = v[0], eventIndex = v[1];
+
+        var endIndex = Math.min(tl.getEvents().length,
+                                eventIndex + initialWindowSize / 2);
+        var startIndex = Math.max(0, endIndex - initialWindowSize);
+        self._start = new TimelineIndex(tl, startIndex - tl.getBaseIndex());
+        self._end = new TimelineIndex(tl, endIndex - tl.getBaseIndex());
+        self._eventCount = endIndex - startIndex;
+    });
+
+    return prom;
+};
+
+/**
+ * Check if this window can be extended
+ *
+ * <p>This returns true if we either have more events, or if we have a
+ * pagination token which means we can paginate in that direction. It does not
+ * necessarily mean that there are more events available in that direction at
+ * this time.
+ *
+ * @param {string} direction   EventTimeline.BACKWARDS to check if we can
+ *   paginate backwards; EventTimeline.FORWARDS to check if we can go forwards
+ *
+ * @return {boolean} true if we can paginate in the given direction
+ */
+TimelineWindow.prototype.canPaginate = function(direction) {
+    var tl;
+    if (direction == EventTimeline.BACKWARDS) {
+        tl = this._start;
+        if (tl.index > tl.minIndex()) { return true; }
+    } else if (direction == EventTimeline.FORWARDS) {
+        tl = this._end;
+        if (tl.index < tl.maxIndex()) { return true; }
+    } else {
+        throw new Error("Invalid direction '" + direction + "'");
+    }
+
+    return tl.timeline.getNeighbouringTimeline(direction) ||
+        tl.timeline.getPaginationToken(direction);
+};
+
+/**
+ * Attempt to extend the window
+ *
+ * @param {string} direction   EventTimeline.BACKWARDS to extend the window
+ *    backwards (towards older events); EventTimeline.FORWARDS to go forwards.
+ *
+ * @param {number} size   number of events to try to extend by. If fewer than this
+ *    number are immediately available, then we return immediately rather than
+ *    making an API call.
+ *
+ * @param {boolean} [makeRequest = true]  whether we should make API calls to
+ *    fetch further events if we don't have any.
```

Comment:
Some clarification on how this interacts with `size` would be great. E.g:
 - `size=7`, timeline has 5, `makeRequest=true` = ??? (the docs on `size` suggest no request will be made since "fewer than this number are immediately available", but I've explicitly set `makeRequest=true`. Which one wins?